src/core/response/mod.rs

```diff
@@ -26,45 +22,46 @@ impl AuthorizationResponse {
             return Ok(Self::Jwt(jwt));
         }
 
-        let flattened = serde_urlencoded::from_bytes::<BTreeMap<String, String>>(bytes)
+        let unencoded = serde_urlencoded::from_bytes::<UnencodedAuthorizationResponse>(bytes)
             .context("failed to construct flat map")?;
-        let map = flattened
-            .into_iter()
-            .map(|(k, v)| {
-                let v = serde_json::from_str::<Value>(&v).unwrap_or(Value::String(v));
-                (k, v)
-            })
-            .collect();
-
-        Ok(Self::Unencoded(UntypedObject(map).try_into()?))
+
+        Ok(Self::Unencoded(unencoded))
     }
 }
 
-#[derive(Debug, Clone)]
-pub struct UnencodedAuthorizationResponse(
-    pub UntypedObject,
-    pub VpToken,
-    pub PresentationSubmission,
-);
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct UnencodedAuthorizationResponse {
+    pub vp_token: VpToken,
+    pub presentation_submission: PresentationSubmission,
+}
 
 impl UnencodedAuthorizationResponse {
     /// Encode the Authorization Response as 'application/x-www-form-urlencoded'.
     pub fn into_x_www_form_urlencoded(self) -> Result<String> {
-        let mut inner = self.0;
-        inner.insert(self.1);
-        inner.insert(self.2);
-        serde_urlencoded::to_string(inner.flatten_for_form()?)
-            .context("failed to encode response as 'application/x-www-form-urlencoded'")
+        let mut map = HashMap::<&str, String>::new();
+        let vp_token =
+            serde_json::to_string(&self.vp_token).context("failed to parse JSON vp token")?;
+        let presentation_submission = serde_json::to_string(&self.presentation_submission)
+            .context("failed to encode presentation_submission as JSON")?;
+
+        map.insert("presentation_submission", presentation_submission);
+        map.insert("vp_token", vp_token);
+
+        let encoded = serde_urlencoded::to_string(&map).context(
+            "failed to encode presentation_submission as 'application/x-www-form-urlencoded'",
+        )?;
+
+        Ok(encoded)
```

Comment:
It works so I'm not going to block it but:
- there is no reason for the serialization to fail, so you can safely `unwrap` instead of returning a `Result`
- you could use a custom `struct` instead of a `HashMap` to avoid memory allocations.

Reply:
commit [ad3974c](https://github.com/spruceid/openid4vp/pull/39/commits/ad3974c5d19c94ecefb4b83427886ff1c480c194) removes the hash map and uses a custom struct here instead.